[export] Constraints API

angelayi opened the pull request:
Wrapper for users to insert constraints into model code.

The constraints will not be maintained in the graph after tracing through make_fx so retracing with dynamo/make_fx will not work. This will be supported after torch._assert supported is implemented. Then we can convert the constrain_range calls to torch._asserts.


cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

tugsbayasgalan commented on torch/_export/constraints.py:
```diff
@@ -0,0 +1,37 @@
+            max = ctx.get_local("max").as_python_constant()
+            constrain_range(node.meta["example_value"], min=min, max=max)
+            ctx.graph().call_function(constrain_range, (node,), {"min": min, "max": max})
+    else:
```
When would this happen?

angelayi replied:
If you just trace with make_fx then this @comptime function doesn't get called so there's no constraint added.

tugsbayasgalan commented on torch/_export/constraints.py:
```diff
@@ -0,0 +1,37 @@
+
+    if min is not None and min <= 2:
+        if not torch._dynamo.is_compiling():
+            warnings.warn("Unable to set min size to be <= 2 because we specialize on 0/1 sizes.")
```
Can you leave a TODO saying we should investigate turning it off for unbacked SymInt

tugsbayasgalan replied:
Hmm maybe we should just error lol. So the behaviour would be if user doesn't provide min/max we just assume it is 2. If they provide, we check if it is less than 2, and if it is, we error.

tugsbayasgalan commented on torch/_export/constraints.py:
```diff
@@ -0,0 +1,37 @@
+from torch.fx.experimental.symbolic_shapes import constrain_range
+
+
+def add_inline_constraint(symbol, min: Optional[int] = None, max: Optional[int] = None):
```
Can you leave TODO here that we want to hide this min/max stuff under DynamicVar or some kind of abstraction that is similar to DynamicDim.

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -75,5 +79,34 @@ def forward(self, x):
+            for i, x in enumerate(roi_batch_splits_nms.to(torch.int32)):
+                b = x.item()
+
+                add_inline_size_constraint(b, min=0)
```
What happens if you call this again, but with different min/max?

angelayi replied:
It'll add both constraints into the graph.

tugsbayasgalan replied:
Hmm should we add a check to make sure it errors if the second constraint doesn't overlap with the previous one? Is it feasible?

angelayi replied:
Hmm, not sure...it'll error when tracing with dynamo/make_fx though.

tugsbayasgalan replied:
Ohh hmm can you add a test case where you add conflicting constrains on a symbol and assert that it errors during tracing? We want to make sure the error message is also clear to user

angelayi replied:
done!

tugsbayasgalan commented on torch/_export/constraints.py:
```diff
@@ -0,0 +1,36 @@
+
+    # TODO: we should investigate turning off 0/1 specialization for unbacked
+    # SymInts
+    assert min >= 2, "Unable to set min size to be <= 2 because we specialize on 0/1 sizes."
```
Can you make this Runtime Error since it is a user error?

avikchaudhuri replied:
Actually I don't like that we have this extra function just for convenience. Maybe simply force min to be the max(2, whatever the user passed in).

avikchaudhuri commented on torch/_export/constraints.py:
```diff
@@ -0,0 +1,36 @@
+            node = ctx.get_local("symbol").as_proxy().node
+            min = ctx.get_local("min").as_python_constant()
+            max = ctx.get_local("max").as_python_constant()
+            constrain_range(node.meta["example_value"], min=min, max=max)
```
Curious, mostly for my education: Why is there a double call here (vs. the one below)? Is `example_value` the concrete value of the symbol? Do we have such a thing for an unbacked symint?

angelayi replied:
Oh sorry, you're right that we don't need the first call -- I forgot that the 2nd line calls the function.

avikchaudhuri commented on torch/_export/constraints.py:
```diff
@@ -0,0 +1,36 @@
+        @comptime
+        def _(ctx):
+            node = ctx.get_local("symbol").as_proxy().node
+            min = ctx.get_local("min").as_python_constant()
```
Shouldn't there be any checking going on here? Since `add_inline_constraint` is a user-facing API, they might make mistakes in passing the wrong types of inputs...

angelayi replied:
I think they should be caught at runtime since the constrain_range function will be called while tracing, or at least be caught by a linter...

avikchaudhuri commented on torch/_export/constraints.py:
```diff
@@ -0,0 +1,36 @@
+
+    return symbol
+
+def add_inline_size_constraint(symbol, min: int = 2, max: Optional[int] = None):
```
I don't see the point of having two very similar functions exposed to the user. Pick one.

angelayi replied:
One is for adding a constraint on a value you know will be used as the size of some other tensor -- this has to be >= 2 and therefore has an extra check. The other is for adding a constraint on just any value which I think can be any integer right?

tugsbayasgalan replied:
Yeah that was the intention (one for size constraints and other for any arbitrary symbol). We have seen a case where user is using the unbacked symint to do data-dependent slicing. For that, i think it is little weird to force the symbol value to be 2 because if user does something like `n - unbacked_val`, the shape can be >= 2 and we don't need to specialize.

avikchaudhuri replied:
OK, may I suggest then that we use different names to make the distinction more explicit. For the first, `constrain_as_value` and the for the second, `constrain_as_size`. The words `add` and `inline` don't provide much information, so take them out.

angelayi replied:
Yup! Those names sound better! Thanks!

avikchaudhuri commented on torch/_export/trace.py:
```diff
@@ -0,0 +1,212 @@
+__all__ = ["do_not_use_experimental_export"]
+
+
+def _aot_capture(mod, flat_args):
```
Does this change have anything to do with inline constraints? Not sure why this is in the same diff, maybe I'm missing something?

angelayi replied:
It has to do with circular imports 😅 

I need to add torch/_export/constraints.py to dynamo's skipfiles so that dynamo knows to trace through these functions. But there was a circular import error in functorch's aot_autograd tests, something like: aot_autograd tests import dynamo --> dynamo imports torch/_export/constraints.py in its skipfiles --> torch/_export/__init__.py imports aot_autograd.

So if I move tracer code to another file (not __init__.py) it'll solve that issue 😋

avikchaudhuri reviewed (changes requested):
Cool! Requesting changes mainly because I'd like a single API.

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -75,5 +79,46 @@ def forward(self, x):
+        res = gm(*inp)
+        self.assertTrue(torchdynamo.utils.same(ref, res))
+
+    @unittest.skip("RuntimeError: Python 3.11+ not yet supported for torch.compile")
```
Can you make it skipIf? I think i saw some utility functions in dynamo about skipping if python 3.11.

tugsbayasgalan commented on torch/_export/constraints.py:
```diff
@@ -0,0 +1,34 @@
+    # TODO: we should investigate turning off 0/1 specialization for unbacked
+    # SymInts
+    if min < 2:
+        raise UserError(
```
same here. You can just use ValueRangeError here.

angelayi commented:
@pytorchbot merge

izaitsevfb commented:
@pytorchbot -revert -m "Breaks internal tests, asked by author to revert"

izaitsevfb commented:
@pytorchbot revert -m "Breaks internal tests, asked by author to revert"

izaitsevfb commented:
@pytorchbot revert -m "Breaks internal tests, asked by author to revert" -c nosignal

angelayi committed:
[export] Constraints API (#98433)

Summary:
Wrapper for users to insert constraints into model code.

The constraints will not be maintained in the graph after tracing through make_fx so retracing with dynamo/make_fx will not work. This will be supported after torch._assert supported is implemented. Then we can convert the constrain_range calls to torch._asserts.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Pull Request resolved: https://github.com/pytorch/pytorch/pull/98433

Differential Revision: D44941579

Pulled By: angelayi

fbshipit-source-id: 606e30153dc54a89b7a1635c69db4eef1d4c5e33

angelayi commented:
@pytorchbot merge
